### ISSUE TITLE
Putting An End to End-to-End: Gradient-Isolated Learning of Representations

### ISSUE THREAD

**Issue by juanmc2005** [juanmc2005/phd-bib#16]
## [Putting An End to End-to-End: Gradient-Isolated Learning of Representations](https://papers.nips.cc/paper/8568-putting-an-end-to-end-to-end-gradient-isolated-learning-of-representations)
*We propose a novel deep learning method for local self-supervised representation learning that does not require labels nor end-to-end backpropagation but exploits the natural order in data instead. Inspired by the observation that biological neural networks appear to learn without backpropagating a global error signal, we split a deep neural network into a stack of gradient-isolated modules. Each module is trained to maximally preserve the information of its inputs using the InfoNCE bound from Oord et al [2018]. Despite this greedy training, we demonstrate that each module improves upon the output of its predecessor, and that the representations created by the top module yield highly competitive results on downstream classification tasks in the audio and visual domain. The proposal enables optimizing modules asynchronously, allowing large-scale distributed training of very deep neural networks on unlabelled datasets.*

### Cite this paper
```bibtex
@incollection{NIPS2019_8568,
	title = {Putting An End to End-to-End: Gradient-Isolated Learning of Representations},
	author = {L\"{o}we, Sindy and O\textquotesingle Connor, Peter and Veeling, Bastiaan},
	booktitle = {Advances in Neural Information Processing Systems 32},
	editor = {H. Wallach and H. Larochelle and A. Beygelzimer and F. d\textquotesingle Alch\'{e}-Buc and E. Fox and R. Garnett},
	pages = {3033--3045},
	year = {2019},
	publisher = {Curran Associates, Inc.},
	url = {http://papers.nips.cc/paper/8568-putting-an-end-to-end-to-end-gradient-isolated-learning-of-representations.pdf}
}
```

**Comment by juanmc2005** [comment]
### Description

This paper introduces a self-supervised learning approach called Greedy InfoMax (GIM) with a particular characteristic: there's no end-to-end backpropagation. The model is split into a subset of consecutive layers that are trained individually with a mutual information maximization objective, very much like CPC.

The technique is motivated by the fact that no end-to-end training seems to exist in the brain, while it has been hypothesized that the brain maximizes information between different neuron layers.
The immediate advantage of this approach is that it unlocks asynchronous pipelining of layers, leading to lower computational costs, as well as reducing the amount of memory needed during training.

They perform experiments in the visual and audio domains and find that their approach achieves competitive performance w.r.t. a CPC baseline (even outperforming it in some cases).

### Approach

The technique is very well described in the paper, the key things to remember are that it closely resembles CPC at each layer. The CPC autoregressive (pooling) layer is removed for vision but kept for audio.

![image](https://user-images.githubusercontent.com/5760780/79115490-3bc3d180-7d86-11ea-9ccc-c91f99bfdfad.png)

### Architecture

They use a ResNet-50 v2 (split into 3 modules) for the vision experiments, and a model consisting of 5 convolutional layers and a recurrent layer for the audio experiments.

### Datasets

They use STL-10 for vision and Librispeech for audio.

### Evaluation

They train a classification layer freezing the network weights and report standard accuracy.

### Results

On vision, GIM outperforms DIM (#20), CPC, and its supervised counterpart on classification accuracy. On audio, results are quite mixed: GIM performs competitively w.r.t. CPC on phone classification accuracy and speaker classification accuracy. However, it performs worse than the supervised counterparts on phone classification, and better than them on speaker classification.
The authors suggest that this is because GIM might concentrate on global representations rather than local.

They also show that memory usage can be reduced by 2.8 (for vision) by splitting the model into 3 modules.

Another interesting aspect is that iterative training (training one module to convergence at a time) seems to hurt final performance. They attribute this to initial noise in representations that might have a regularizing effect.

Finally, an ablation study on audio shows that the autoregressive layer is of little help for speaker classification, while it makes a big difference on phone classification.

### Takeaways

This is a really interesting paper and they introduce a really attractive direction for self-supervised learning in general, showing that there's no immediate advantage of doing end-to-end training.

Moreover, modular models like this allow for incremental model training (instead of fine-tuning), lower memory footprint and better parallelization via pipelining.

It is an interesting approach to representation learning, as it maximizes mutual information at each module, and it could have applications for continual learning, by allocating independent neural resources when using dynamic expansion.
